### PR TITLE
refactor to use std::shared_ptr with peer_connection, core and tests

### PR DESCRIPTION
**Pull request by aldenml**

### CODE REVIEW

File: bindings/c/library.cpp [hunk 1]
```diff
@@ -60,7 +60,7 @@ namespace
 	int add_handle(libtorrent::torrent_handle const& h)
 	{
 		std::vector<libtorrent::torrent_handle>::iterator i = std::find_if(handles.begin()
-			, handles.end(), !boost::bind(&libtorrent::torrent_handle::is_valid, _1));
+			, handles.end(), !std::bind(&libtorrent::torrent_handle::is_valid, _1));
```
Comment (arvidn) [inline bindings/c/library.cpp]: I'm actually a bit surprised this works. are you sure it means the same thing as with boost::bind? boost has some lambda capabilities which did not make it into the standard, and I would expect this to be one of them.

Comment (aldenml) [inline bindings/c/library.cpp]: this is the broken c binding, it does not compile and it's not usable

Comment (aldenml) [inline bindings/c/library.cpp]: but I saw what you say in the python binding, I needed to used `boost::bind` there in a few places

Comment (arvidn) [inline bindings/c/library.cpp]: would you mind replacing this bind() with a lambda expression? [](libtorrent::torrent_handle const& h) { return !h.is_valid(); }

**Comment by arvidn** [comment]
lgtm (but see comment)

File: bindings/c/library.cpp (hunk 1 above)
Comment (arvidn) [inline bindings/c/library.cpp]: ah, so you could use a lambda there then

**Comment by aldenml** [comment]
do you mind if I do it in another PR?

**Pull request merged**